In [ ]:
Name:

Disease_Prediction

Title: Disease_Prediction Using Machine Learning

 Title: Disease Prediction Using Machine Learning
 
 The project involves the application of machine learning techniques to predict diseases based on available data from the datasets.
 Predicting Stomach Pain according to disease_prediction using Machine Learning.
 
Problem Definition: this project aims to predict the occurrence of stomach pain based on various features using classification techniques.

Classification,problem:random forest classifier workflow for classifying Disease Prediction 

Problem Statement: To predict whether a patient has stomach pain based on various medical features available in the dataset  
(To develop a model that can predict whether a person has stomach pain based on certain symptoms and features.)

In [157]:
#import the standard library
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt

In [158]:
# Load data
Testing=pd.read_csv("Testing.csv")
Testing

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0         1          1                     1                    0          0   
1         0          0                     0                    1          1   
2         0          0                     0                    0          0   
3         1          0                     0                    0          0   
4         1          1                     0                    0          0   
5         0          0                     0                    0          0   
6         0          0                     0                    0          0   
7         0          0                     0                    0          0   
8         0          0                     0                    0          0   
9         0          0                     0                    0          0   
10        0          0                     0                    0          0   
11        0          0                     0                    0          0   
12        0          0                     0                    0          0   
13        0          0                     0                    0          0   
14        1          0                     0                    0          0   
15        0          0                     0                    0          0   
16        1          1                     0                    0          0   
17        0          1                     0                    0          0   
18        0          0                     0                    0          0   
19        0          0                     0                    0          0   
20        1          0                     0                    0          0   
21        0          0                     0                    0          0   
22        0          0                     0                    0          0   
23        0          0                     0                    0          0   
24        0          0                     0                    0          0   
25        0          0                     0                    0          0   
26        0          0                     0                    1          0   
27        0          0                     0                    0          0   
28        0          0                     0                    0          0   
29        0          0                     0                    0          0   
30        0          0                     0                    0          0   
31        0          0                     0                    0          0   
32        0          0                     0                    0          0   
33        0          0                     0                    0          0   
34        0          0                     0                    0          0   
35        0          0                     0                    0          0   
36        0          0                     0                    0          0   
37        0          1                     0                    0          0   
38        0          0                     0                    0          0   
39        0          1                     0                    0          0   
40        0          1                     0                    0          0   
41        1          1                     0                    0          0   

    chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  \
0        0           0             0        0                 0  ...   
1        1           0             0        0                 0  ...   
2        0           0             1        1                 1  ...   
3        0           0             0        0                 0  ...   
4        0           0             1        0                 0  ...   
5        0           0             0        0                 0  ...   
6        0           0             0        0              

**About The Dataset**

For our disease classification example, we have a dataset with 133 columns and 42 rows.
Each row corresponds to a unique case, and each column represents a different feature. The features are symptoms, and the target variable is the stomach_pain, indicating the predicted outcome or class for each case.

The 133 columns in our dataset represent different symptoms observed in patients.
    These symptoms could range from common indicators like fever and cough to more specific ones depending on the nature of the diseases under consideration. The features play a crucial role 
in the classification process as they are the input variables used by the algorithm to make predictions.

In [159]:
###Load Data Loading the datasets

df = pd.read_csv("datasets.csv")

#identify the feature and the label

x = df.drop("stomach_pain",axis=1)     # Features
y = df["stomach_pain"]                  # Labels
print(y.head())
print(x.head())

0    0
1    0
2    1
3    0
4    0
Name: stomach_pain, dtype: int64
   itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        1          0                     0                    0          0   
2        1          1                     0                    0          0   
3        0          0                     0                    0          0   
4        0          0                     0                    0          0   

   chills  joint_pain  acidity  ulcers_on_tongue  yellow_crust_ooze  \
0       0           0        0                 0                  0   
1       0           0        0                 0                  0   
2       0           0        0                 0                  0   
3       0           0        0                 0                  0   
4       0           0        0                 0                  0   

             prognosis  
0    

	Features and Label

Features are the input variables used to make predictions.

Labels are the target variables that we want to predict.

   The features are all columns except "stomach_pain", and the label is the "stomach_pain" column itself.

Data Dictionary

fever: Binary feature indicating if the person has a fever.

nausea: Binary feature indicating if the person is experiencing nausea.

fatigue: Binary feature indicating if the person is feeling fatigue.

cough: Binary feature indicating if the person has a cough.

diarrhea: Binary feature indicating if the person has diarrhea.

stomach_pain: Binary target label indicating if the person has stomach pain.

Data

Sourcing : Collecting medical data from reliable sources or healthcare institutions.

The dataset is sourced from CSV files named "Testing.csv".

Defining Different Parameters:Identifying key features such as age, gender, medical history, lifestyle factors, genetic predispositions, etc.

Features such as fever, nausea, fatigue, etc., and a target label "stomach_pain".

Consultation:Talking to experts: Consulting medical professionals to understand the relevance and significance of different parameters in disease prediction.

Discussions with healthcare professionals to understand the features and their relevance to stomach pain.

In [160]:
from sklearn.model_selection import train_test_split

In [161]:
# Split data into training and testing sets
x_train ,x_test,y_train,y_test=train_test_split(x,y,test_size=0.30)

Training and Test Split

   The dataset is split into training and test sets to evaluate the model's performance on unseen data.

   The training set is used to train the model, and the test set is used to evaluate it.

In [162]:
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((14, 11), (7, 11), (14,), (7,))

In [163]:
#model train 
#model choosing 
from sklearn.ensemble import RandomForestClassifier


Model Choices

   Different models can be chosen based on the problem at hand.
       
   In your project of disease prediction, you are using "RandomForestClassifier".

In [164]:
#model created
# Define a RandomForestClassifier
clf=RandomForestClassifier()

In [165]:
# Define a parameter grid for hyperparameter tuning
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10]
}

In [166]:
# Import GridSearchCV from sklearn.model_selection
from sklearn.model_selection import GridSearchCV

In [167]:
#learn the model creating object by calling the model
clf.fit(x_train,y_train)

ValueError: could not convert string to float: 'Hepatitis D'

In [168]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   itching               21 non-null     int64 
 1   skin_rash             21 non-null     int64 
 2   nodal_skin_eruptions  21 non-null     int64 
 3   continuous_sneezing   21 non-null     int64 
 4   shivering             21 non-null     int64 
 5   chills                21 non-null     int64 
 6   joint_pain            21 non-null     int64 
 7   stomach_pain          21 non-null     int64 
 8   acidity               21 non-null     int64 
 9   ulcers_on_tongue      21 non-null     int64 
 10  yellow_crust_ooze     21 non-null     int64 
 11  prognosis             21 non-null     object
dtypes: int64(11), object(1)
memory usage: 2.1+ KB


Data Exploration (Exploratory Data Analysis or EDA)

Exploratory Data Analysis (EDA) is a critical step in understanding your dataset and preparing it for modeling

What question(s) are you trying to solve

   The main question you are trying to solve is: Can we predict if a patient has stomach pain based on other features in the dataset
   

What kind of data do you have and how do you treat different types?

   You have a dataset that includes various features (predictors) and a target variable ("stomach_pain").

   Identify the data types of each feature (numerical, categorical, etc.) and decide how to handle them.
       
   Categorical data might need encoding (e.g., using OneHotEncoder), while numerical data might need scaling or normalization.
       
To change or convert string to float and ignore or remove unknown

In [169]:

from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
cat_feature = ["prognosis"]
onehot = OneHotEncoder(handle_unknown="ignore")
transformer = ColumnTransformer([("onehot", onehot, cat_feature)], remainder = "passthrough")
encoded_x = transformer.fit_transform(x_train)
encoded_x_test = transformer.transform(x_test)
encoded_x


array([[0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
        0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0.,
        0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0.,
        0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1.,
        0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.,
        0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
        0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.,
        0., 0., 1., 1., 0., 0., 0.],
       [0., 1., 0.,

Preparing the Tools: list belows to use the tools

Importing the necessary libraries:  jupyter notebook

import numpy as np

import pandas as pd

import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, GridSearchCV

import seaborn as sns

from sklearn.model_selection import train_test_split

from sklearn.ensemble import RandomForestClassifier

from sklearn.preprocessing import OneHotEncoder

from sklearn.compose import ColumnTransformer

from sklearn.metrics import accuracy_score, classification_report

import pickle

In [170]:

encoded_x_df = pd.DataFrame(encoded_x)
encoded_test_df = pd.DataFrame(encoded_x_test)
encoded_x_df


0    1    2    3    4    5    6    7    8    9   ...   13   14   15   16  \
0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0   
2   0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0  0.0   
3   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  1.0  0.0  0.0   
5   0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  1.0  1.0  0.0   
6   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
7   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
8   0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0  0.0   
9   0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0  0.0   
11  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  1.0   
12  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
13  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

     17   18   19   20   21   22  
0   0.0  0.0  1.0  0.0  0.0  0.0  
1   0.0  1.0  0.0  0.0  0.0  0.0  
2   0.0  0.0  0.0  0.0  0.0  0.0  
3   0.0  0.0  0.0  1.0  0.0  0.0  
4   0.0  0.0  0.0  0.0  0.0  0.0  
5   0.0  0.0  0.0  0.0  0.0  0.0  
6   0.0  1.0  0.0  0.0  0.0  0.0  
7   0.0  1.0  0.0  0.0  0.0  0.0  
8   0.0  1.0  1.0  0.0  0.0  0.0  
9   0.0  0.0  0.0  0.0  0.0  0.0  
10  0.0  0.0  1.0  0.0  0.0  0.0  
11  0.0  1.0  0.0  0.0  0.0  0.0  
12  0.0  0.0  0.0  0.0  0.0  0.0  
13  0.0  0.0  0.0  0.0  0.0  0.0  

[14 rows x 23 columns]

In [171]:
###Missing Data Checking for missing values
df.isnull().sum()

itching                 0
skin_rash               0
nodal_skin_eruptions    0
continuous_sneezing     0
shivering               0
chills                  0
joint_pain              0
stomach_pain            0
acidity                 0
ulcers_on_tongue        0
yellow_crust_ooze       0
prognosis               0
dtype: int64

In [172]:
#model train 
#model choosing 
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()



In [173]:
clf.fit(encoded_x,y_train)

RandomForestClassifier()

In [174]:
importances = clf.feature_importances_
feature_names = transformer.get_feature_names_out()
feature_importances = pd.DataFrame(importances, index=feature_names, columns=['Importance'])
print(feature_importances.sort_values(by='Importance', ascending=False))

                                       Importance
onehot__prognosis_Alcoholic hepatitis         0.0
onehot__prognosis_Tuberculosis                0.0
remainder__ulcers_on_tongue                   0.0
remainder__acidity                            0.0
remainder__joint_pain                         0.0
remainder__chills                             0.0
remainder__shivering                          0.0
remainder__continuous_sneezing                0.0
remainder__nodal_skin_eruptions               0.0
remainder__skin_rash                          0.0
remainder__itching                            0.0
onehot__prognosis_Pneumonia                   0.0
onehot__prognosis_Arthritis                   0.0
onehot__prognosis_Migraine                    0.0
onehot__prognosis_Malaria                     0.0
onehot__prognosis_Hepatitis E                 0.0
onehot__prognosis_Hepatitis D                 0.0
onehot__prognosis_Hepatitis C                 0.0
onehot__prognosis_Hepatitis B                 0.0


In [175]:
pred=clf.predict(encoded_x_test)

In [176]:
print(x_test.head())

    itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
1         1          0                     0                    0          0   
10        0          0                     0                    0          0   
4         0          0                     0                    0          0   
9         0          0                     0                    0          0   
5         0          0                     0                    0          0   

    chills  joint_pain  acidity  ulcers_on_tongue  yellow_crust_ooze  \
1        0           0        0                 0                  0   
10       0           1        0                 0                  0   
4        0           0        0                 0                  0   
9        1           0        0                 0                  0   
5        0           0        0                 0                  0   

              prognosis  
1   Chronic cholestasis  
10          hepatitis A  
4       

In [177]:
pred

array([0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [178]:
print(y_test)
print(pred)

1     0
10    0
4     0
9     0
5     0
2     1
3     0
Name: stomach_pain, dtype: int64
[0 0 0 0 0 0 0]


In [179]:
##Evaluation
print(f"{clf.score(encoded_x_test,y_test)*100:.2f}%")

85.71%


In [180]:
#expremenation
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [181]:
for i in range(100,200,10):
    print(f"values of estimators ={i}")
    model =RandomForestClassifier(n_estimators=i)
    model.fit(encoded_x,y_train)
    print(f"{model.score(encoded_x_test,y_test)*100:.2f}%")

values of estimators =100
85.71%
values of estimators =110
85.71%
values of estimators =120
85.71%
values of estimators =130
85.71%
values of estimators =140
85.71%
values of estimators =150
85.71%
values of estimators =160
85.71%
values of estimators =170
85.71%
values of estimators =180
85.71%
values of estimators =190
85.71%


In [182]:
np.random.seed(42)
clf=RandomForestClassifier(n_estimators=120).fit(encoded_x,y_train)

In [183]:
print(f"{clf.score(encoded_x_test,y_test)*100:.2f}%")

85.71%


After experimentation, you found that the model achieves an accuracy of 85.71% when the number of estimators is set to 120 or all are the same.

Iterates over different values of n_estimators from 100 to 190 with a step of 10. For each value, it trains a RandomForestClassifier and calculates the accuracy on the test set. 
Finally, it selects the value of n_estimators with the highest accuracy (120 in this case) and trains the final model with that value.

Did you achieve your evaluation if not? yes to achieve greater than 80%

In [184]:
from sklearn.metrics import accuracy_score, classification_report

In [185]:
predictions = clf.predict(encoded_x_test)

In [186]:
##Evaluation

### Making Predictions and Evaluating the Model

pred = clf.predict(encoded_x_test)
print(f"Accuracy: {accuracy_score(y_test, pred)*100:.2f}%")
print(classification_report(y_test, pred))

Accuracy: 85.71%
              precision    recall  f1-score   support

           0       0.86      1.00      0.92         6
           1       0.00      0.00      0.00         1

    accuracy                           0.86         7
   macro avg       0.43      0.50      0.46         7
weighted avg       0.73      0.86      0.79         7



C:\Users\AMAN_CoSc\anaconda4\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\AMAN_CoSc\anaconda4\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\AMAN_CoSc\anaconda4\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


"accuracy_score(y_test, predictions)" calculates the accuracy by comparing the true labels (y_test) with the predicted labels (predictions).

The accuracy is then printed as a percentage.

Precision: The ratio of correctly predicted positive observations to the total predicted positives.

Recall (Sensitivity): The ratio of correctly predicted positive observations to all observations in the actual class.

F1-Score: The weighted average of Precision and Recall.

Support: The number of actual occurrences of each class in the specified dataset.

Model Comparison

   It's often useful to compare different models to see which performs best on your data.
   
   You can use metrics like accuracy, precision, recall, and F1-score for comparison.


In [187]:
import pickle

Save the Model(stomach_pain) by using pickle import to save the model of your project

The "pickle.dump" function serializes the trained model ("clf") and saves it to a file named "stomach_pain_model.pkl".

In [188]:
pickle.dump(clf,open("Testing.pkl","wb"))

Exporting the Model based on datasets such as Testing 

In [189]:
#load the model
loaded_model= pickle.load(open("Testing.pkl","rb"))

Loading the model  to use : The "pickle.load" function deserializes the model from the file and loads it back into a variable ("pickle.load").

In [190]:
loaded_model.predict(encoded_x_test)

array([0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [191]:
print(f"{clf.score(encoded_x_test,y_test)*100:.2f}%")

85.71%


Testing the Loaded Model

   The loaded model is used to make predictions on the test data ("encoded_x_test").

   The predictions are compared with the actual labels ("y_test") to evaluate the performance of the loaded model using accuracy and classification report.